onnxruntime/core/common/threadpool.cc

```diff
@@ -21,12 +21,222 @@ limitations under the License.
 #include "core/common/eigen_common_wrapper.h"
 #include "core/platform/EigenNonBlockingThreadPool.h"
 #include "core/platform/ort_mutex.h"
+#ifdef _WIN32
+#include "processthreadsapi.h"
+#include <codecvt>
+#include <locale>
+#elif defined(__APPLE__)
+#include <cpuid.h>
+#else
+#include <sched.h>
+#endif
 
 namespace onnxruntime {
 
 namespace concurrency {
 
-// A sharded loop counter distributes loop iterations between a set of worker threads.  The iteration space of
+ThreadPoolProfiler::ThreadPoolProfiler(int num_threads, const CHAR_TYPE* threal_pool_name) : 
+    num_threads_(num_threads) {
+  child_thread_stats_.assign(num_threads, {});
+  if (threal_pool_name) {
+#ifdef _WIN32
+    using convert_type = std::codecvt_utf8<wchar_t>;
+    std::wstring_convert<convert_type, wchar_t> converter;
+    threal_pool_name_ = converter.to_bytes(threal_pool_name);
+#else
+    threal_pool_name_ = threal_pool_name;
+#endif
+  } else {
+    threal_pool_name_ = "unnamed_thread_pool";
+  }
+}
+
+ThreadPoolProfiler::~ThreadPoolProfiler() {
+  enabled_ = false;
+}
+
+void ThreadPoolProfiler::Start() {
+  enabled_ = true;
+}
+
+ThreadPoolProfiler::MainThreadStat& ThreadPoolProfiler::GetMainThreadStat() {
+  static thread_local std::unique_ptr<MainThreadStat> stat;
+  if (!stat) {
+    stat.reset(new MainThreadStat());
+  }
+  return *stat;
+}
+
+std::string ThreadPoolProfiler::Stop() {
+  ORT_ENFORCE(enabled_, "Profiler not started yet");
+  std::stringstream ss;
+  ss << "{\"main_thread\": {"
+     << "\"thread_pool_name\": \""
+     << threal_pool_name_ << "\", "
+     << GetMainThreadStat().Reset()
+     << "}, \"sub_threads\": {"
+     << DumpChildThreadStat()
+     << "}}";
+  return ss.str();
+}
+
+void ThreadPoolProfiler::LogCoreAndBlock(std::ptrdiff_t block_size) {
```

Comment:
Ref the comment above on passing in block_size to RunInParallel just in order to log it: one option is to split these pieces apart, and keep the LogCore calls from within the inner loops in EigenNonBlockingThreadPool, and then have a separate LogBlockSize function that can be called from within threadpool.cc to log the block size.

Reply:
Like replied above, block_size and profiler exist in two different realm - the former owned by thread pool while the latter is a member of thread pool impl. Even if we have a separate "LogBlockSize" for profiler, how to enable the thread pool of calling it? One plain approach it to add one more interface into thread pool impl - "LogBlockSize" which calls into the corresponding function of profiler, but is this neater than the present approach?